arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -393,25 +395,75 @@ static bool sync_shadow_pasid_dir_entry(struct pgt_sync_data *sdata)
 	return false;
 }
 
+static struct pkvm_ptdev *iommu_find_ptdev(struct pkvm_iommu *iommu, u16 bdf, u32 pasid)
+{
+	struct pkvm_ptdev *p;
+
+	list_for_each_entry(p, &iommu->ptdev_head, iommu_node) {
+		if (match_ptdev(p, bdf, pasid))
+			return p;
+	}
+
+	return NULL;
+}
+
+static struct pkvm_ptdev *iommu_add_ptdev(struct pkvm_iommu *iommu, u16 bdf, u32 pasid)
+{
+	struct pkvm_ptdev *ptdev = pkvm_get_ptdev(bdf, pasid);
+
+	if (!ptdev)
+		return NULL;
+
+	list_add_tail(&ptdev->iommu_node, &iommu->ptdev_head);
+	return ptdev;
+}
+
+static void iommu_del_ptdev(struct pkvm_iommu *iommu, struct pkvm_ptdev *ptdev)
+{
+	list_del_init(&ptdev->iommu_node);
+	pkvm_put_ptdev(ptdev);
+}
+
 /* sync pasid table entry when guest_ptep valid, otherwise un-present it */
 static bool sync_shadow_pasid_table_entry(struct pgt_sync_data *sdata)
 {
+	u16 bdf = sdata->vaddr >> DEVFN_SHIFT;
+	u32 pasid = sdata->vaddr & ((1UL << MAX_NR_PASID_BITS) - 1);
+	struct pkvm_iommu *iommu = pgt_to_pkvm_iommu(sdata->spgt);
+	struct pkvm_ptdev *ptdev = iommu_find_ptdev(iommu, bdf, pasid);
 	struct pasid_entry *shadow_pte = sdata->shadow_ptep, tmp_pte = {0};
 	struct pasid_entry *guest_pte;
+	bool synced = false;
 	u64 type, aw;
 
+	if (!ptdev) {
+		ptdev = iommu_add_ptdev(iommu, bdf, pasid);
+		if (!ptdev)
+			return false;
```

Comment:
The shadow_pte is set at the end of this function, so it is not configured when returning. Returning with shadow_pte un-changed seems fine to me. Wondering if you have seen any issue if not zeroing shadow_pte?

Reply:
Not really, it works fine in my tests. I was just wondering if there may be cases when we previously configured this shadow_pte with some non-zero value, and now we need to update it with some other value but we encounter this iommu_add_ptdev() failure or some other failure, and it is unsafe or insecure to just leave this shadow_pte present with the old value. But at the moment I don't know any real examples of such cases.

BTW in the case of iommu_audit_did() failure below, I'm actually observing the opposite: the existing logic works fine, and if I change it to do zeroing of shadow_pte instead of just returning false, then it becomes broken: if I run a protected VM with 2 devices attached, the first device doesn't work. I've found the reason is that iommu_audit_did() unexpectedly fails for the 1st device although it is already attached (so if shadow_pte is zeroed after this failure, it means that we cancel the previous successful iommu setup for this device, so its DMA won't work). If I move iommu_audit_did() check before pkvm_setup_ptdev_did() as suggested by Jason in [3f201ad](https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed#diff-a39c6f1403510a3b05fd7c4e40c5b8c34f1361c3069a15b733376d4c489e479bR453), it fixes this unexpected iommu_audit_did() failure, and then zeroing of shadow_pte doesn't cause a problem anymore. Could you take a look at the discussion in [3f201ad](https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed#diff-a39c6f1403510a3b05fd7c4e40c5b8c34f1361c3069a15b733376d4c489e479bR453)?

Reply:
Yes, replied in the discussion in [3f201ad](https://github.com/intel-staging/pKVM-IA/pull/17/commits/3f201ad27e15b6ece8738070f61dbe9bf29c2aed#diff-a39c6f1403510a3b05fd7c4e40c5b8c34f1361c3069a15b733376d4c489e479bR453)?

Reply:
Regarding iommu_add_ptdev() failure, as Jason noted, it rather indicates an internal problem in pKVM, so maybe we should just PKVM_ASSERT in this case. The same is probably also true for is_pgt_ops_ept() failure.

Reply:
Yes, agree.